### ISSUE TITLE
Choose modules to structure according to Azure SDK not ARM Templates

### ISSUE THREAD

**Issue by mikhailshilkov** [pulumi/pulumi-azure-native#690]
At the moment, Azure-Native puts resources into modules that are named after ARM URL resource providers. Some of those names are outdated or confusing, and some resources clash because they end up in the same large module (like `Network`, e.g. #583).

In many aspects, Azure-Native follows Azure SDKs more closely than ARM Templates: naming resources after operations, flattening properties, renaming properties to SDK names. It would be logical to name modules similar to Azure SDKs too.

For example (old -> new):

- Network -> split out TrafficManager, PrivateDNS, FrontDoor, DNS
- DocumentDB -> Cosmos DB
- DBforPostgreSQL -> PostgreSQL
- DBforMySQL -> MySQL
- Cache -> split to Redis and RedisEnterprise
- Insights -> split to Monitor and ApplicationInsights
- Devices -> IoTHub
- StorSimple -> split to StorSimple1200Series and StorSimple8000Series

etc.

Even, [ARM template docs](https://docs.microsoft.com/en-us/azure/templates/microsoft.documentdb/cassandraclusters?tabs=json) use some of those names in the menu structure: Cosmos DB, DNS, FrontDoor etc.

Every SDK has its own set of mappings, done mostly with `readme.<language>.md` files. So, mappings for different SDKs differ.

One possibility is to use top-level folder names in the azure-rest-api-spec repository folder structure, which gives unambiguous mapping. They are all lowercase so we'd have to map to PascalCase manually.

Obviously, any module naming change is largely breaking.

I'm curious to learn other opinions about this potential change.

**Comment by mikhailshilkov** [comment]
Here is a potential list of changes:

```
Insights->ApplicationInsights
Portal->CloudShell
DocumentDB->CosmosDB
Devices->DeviceProvisioningServices
DevTestLab->DevTestLabs
HybridData->HybridDataManager
Kubernetes->HybridKubernetes
VirtualMachineImages->ImageBuilder
Devices->IotHub
Management->ManagementGroups
DBforMariaDB->MariaDB
Media->MediaServices
OffAzure->Migrate
Migrate->MigrateProjects
Insights->Monitor
ManagedIdentity->Msi
DBforMySQL->MySql
DBforPostgreSQL->PostgreSql
DBforPostgreSQL->PostgreSqlHsc
PowerBI->PowerBIEmbedded
RecoveryServices->RecoveryServicesBackup
RecoveryServices->RecoveryServicesSiteRecovery
Cache->Redis
Cache->RedisEnterprise
Capacity->Reservations
Migrate->ResourceMover
SignalRService->SignalR
StorSimple->StorSimple1200Series
StorSimple->Storsimple8000Series
ImportExport->StorageImportExport
AVS->VMWare
```

Also, some resources move from one module to another:
```
Network->TrafficManager
Network->Frontdoor
Network->Dns
Network->PrivateDns
ContainerService->Compute
Aad->DomainServices
AadIam->AzureActiveDirectory
AzureActiveDirectory->CPIM
Authorization->Resources
Features->Resources
Solutions->Resources
OperationalInsights->ServiceMap
CertificateRegistration->Web
DomainRegistration->Web
```

**Comment by headconnect** [comment]
Sounds good to me - though my test is that it really does have to simplify mental mapping of navigation between various azure resources.
I'm wondering if it wouldn't be pertinent to, along with these proposed changes, introduce a simpler top-level structure to differentiate between core offerings.  Since you're pushing ContainerService into compute, perhaps the same should be done with ContainerInstance, HybridKubernetes (see? I'm using new names already ;)), possibly a few others. (windowsesu, windowsiot, vmwarecloudsimple, "ImageBuilder"..)

For me - maybe identity-related things under `identity`, like all IAM, AzureAD, managedidentity, etc.  `resources` isn't really an optimal name for what it contains - though it's obvious when you're looking for `resourcegroups` and suchlike. 

Basically, what I'm saying is.. this is hard ;) 

I welcome such changes as they make sense as presented, and are not massive in impact. I'm definitely not experienced enough to know what's reasonable, only what "feels right" - so as I'm _that_ subjective, I'm not going to try to impart any wisdom here, only a very lengthy thumbs-up

**Comment by mikhailshilkov** [comment]
@headconnect Thank you for the feedback! Naming is indeed hard.

> you're pushing ContainerService into compute,

I think it's just one legacy-ish resource that goes from ContainerService to Compute, the rest remains the same.

The proposal above is still ~automatically produced from the Open API specs repository, it's just that we pick a different, more nuanced way to determine the name. We still want to do as little naming manually as possible.

**Comment by imod** [comment]
I'm not 100% sure whether I like it or not...
On one side, I do think the names are very strange and I had issues to find the correct types too. But on the other side, after a while I found a way which really made it a lot simpler to find the correct type: by looking at the URLs of the azure portal...
e.g. take this one, while looking at a "Log Analytics Workspace": 
https://portal.azure.com/#@mycomp.com/resource/subscriptions/XXXX-XXXXXX-XXXXX/resourceGroups/mygroup/providers/Microsoft.OperationalInsights/workspaces/mylogs/Overview it clearly points my to `Microsoft.OperationalInsights` 
The same is true when looking at the resources at https://resources.azure.com -> the names are what I use in my code too.

That sad I would rely hate to have to migrate my existing stacks again - if the old names are preconfigured as alias and https://github.com/pulumi/pulumi/issues/6570 is solved, then I guess this is ok.

**Comment by yarinm** [comment]
Will this affect existing stacks' state? This change has to be backwards compatible..

On another note, I think it's better to be aligned with the azure API and if an object is called there DocumentDB we should use that as well. It makes things much easier when trying to understand valid parameter values and if you are already using the azure SDK you will be familiar with the same terminology. I think the current naming is much better.

**Comment by mikhailshilkov** [comment]
> Will this affect existing stacks' state? This change has to be backwards compatible..

We will provide aliases, so it will work in most cases. (https://github.com/pulumi/pulumi/issues/6570 being a counter-example)

> if you are already using the azure SDK you will be familiar with the same terminology

The problem is that literally every SDK is different:

- Go: `cosmos-db/mgmt/2020-04-01-preview/documentdb` (so, both...)
- Python has both `azure-mgmt-cosmosdb` and `azure-mgmt-documentdb`
- Node.js: `cosmos`
- CLI: `cosmosdb`
- .NET: Cosmos
- ARM Templates: `DocumentDB` under the title `Cosmos DB` in the docs

And the combination for every service is different.

**Comment by ie-zero** [comment]
Following the comment from @imod I constantly have the same issue i.e. to find which Pulumi resource maps to which REST call / ARM template. As the top level resources have, or at least I assume they have, a 1-to-1 relation with the REST endpoints would it be possible to automatically generate the documentation linking the two together?

**Comment by mikhailshilkov** [comment]
@ie-zero Yes, that's a good suggestion. We can definitely link from our resources to ARM types - I'm not sure how to deal with the other way around, which seems more useful.

Can you give an example of a comment as you would like to see on a resource that wasn't trivial to you?

**Comment by cleverguy25** [comment]
Added to epic https://github.com/pulumi/home/issues/3552

**Comment by danielrbradley** [comment]
Modules where we're not able to select a single default version beause of the overlap from multiple folders:
- Compute
- Insights
- Media
- Migrate
- Network
- OperationalInsights
- PolicyInsights
- PowerBI
- Resources
- Security
- Web

**Comment by danielrbradley** [comment]
Reviewing the suggested possible renamed above gives us the following potential renames:

- ~`Aad->DomainServices`~ This is actually now called "Azure Entra ID" so this rename is just confusing
- ~`AadIam->AzureActiveDirectory`~ See above
- ~`Authorization->Resources`~ Don't want to combine modules. Also, this is split into both `resources` and `authorization` folders.
- `AVS->VMWare` ❓  Maybe: it's the better name, but perhaps quite niche and searching "AVS" finds the right docs.
- ~`AzureActiveDirectory->CPIM`~ Not a better name
- `Cache->Redis` ✅  Yes, better
- `Cache->RedisEnterprise` ✅ Yes, better
- `Capacity->Reservations` ❓  Maybe: it does align better with docs.
- ~`CertificateRegistration->Web`~ Don't want more in the Web module - it's already large
- ~`ContainerService->Compute`~ This was deprecated and moved to its own containerservice folder.
- ~`DBforMariaDB->MariaDB`~ - not worth breaking change
- ~`DBforMySQL->MySql`~ - not worth breaking change
- ~`DBforPostgreSQL->PostgreSql`~ - not worth breaking change
- `DBforPostgreSQL->PostgreSqlHsc` ❓  Maybe for better separation
- `Devices->DeviceProvisioningServices` ✅ yes: could be a good split
- `Devices->IotHub` ✅ yes: split and rename with above
- ~`DevTestLab->DevTestLabs`~ Not a significant difference
- `DocumentDB->CosmosDB` ✅ yes, much less confusing name
- ~`DomainRegistration->Web`~ Better to keep separate
- ~`Features->Resources`~ Better to keep separate
- ~`HybridData->HybridDataManager`~ Not a significant name difference
- ~`ImportExport->StorageImportExport`~ Not a significant name difference
- `Insights->ApplicationInsights` ✅ Yes: Align to docs
- `Insights->Monitor` ✅ Yes: split from above & align with docs
- ~`Kubernetes->HybridKubernetes`~ Not a significant name difference
- ~`ManagedIdentity->Msi`~ Docs call this Managed Identities
- ~`Management->ManagementGroups`~ Not a significant change
- ~`Media->MediaServices`~ Not a significant change
- `Migrate->MigrateProjects` ❓ Maybe: Only 2 resources from 2018 version
- `Migrate->ResourceMover` ❓  Maybe: This seems legit as migrate is a different tool (see below) but is only 2 resources so not a big issue
- ~`OffAzure->Migrate`~ Better keeping stuff separate for conflicts
- ~`Migrate->Migrate`~ When the namespace and folder are "migrate", it's really migrate, which we already use.
- `Network->Dns` ✅  Done
- `Network->Frontdoor` ✅ Done
- `Network->PrivateDns` ✅ Done
- `Network->TrafficManager` ✅ Done
- ~`OperationalInsights->ServiceMap`~ Service Map will be retired on 30 September 2025
- `Portal->CloudShell` ❓  Maybe: only 4 resources, but this is a better, more searchable name
- ~`PowerBI->PowerBIEmbedded`~ No: Only 1 resource and a similar enough name
- ~`RecoveryServices->RecoveryServicesBackup`~ - versions seem to be released in sync. Docs group all recovery services together
- ~`RecoveryServices->RecoveryServicesSiteRecovery`~ see above
- ~`SignalRService->SignalR`~ Not significant, WebPubSub already split out
- ~`Solutions->Resources`~ Better not to combine
- ~`StorSimple->StorSimple1200Series`~ Doesn't exist
- ~`StorSimple->Storsimple8000Series`~ Also, out of support. Perhaps worth removing.
- ~`VirtualMachineImages->ImageBuilder`~ Quite niche, not significant benefit. Actual docs name is "Azure VM Image Builder".

**Comment by mjeffryes** [comment]
can we close this as done?

**Comment by thomas11** [comment]
The important candidates are renamed in #3918.
